cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterRebalanceAssemblyOperator.java

```diff
@@ -0,0 +1,546 @@
+        RebalanceOptions.RebalanceOptionsBuilder rebalanceOptionsBuilder = new RebalanceOptions.RebalanceOptionsBuilder();
+        if (clusterRebalance.getSpec().getGoals() != null) {
+            rebalanceOptionsBuilder.withGoals(clusterRebalance.getSpec().getGoals());
+        }
+        Future<KafkaClusterRebalanceStatus> nextStateFuture;
+        log.info("{}: Rebalance action from state [{}]", reconciliation, currentState);
+        switch (currentState) {
+            case New:
+                nextStateFuture = onNew(apiClient, rebalanceOptionsBuilder);
+                break;
+            case PendingProposal:
+                nextStateFuture = onPendingProposal(apiClient, clusterRebalance, rebalanceAnnotation);
+                break;
+            case ProposalReady:
+                nextStateFuture = onProposalReady(apiClient, clusterRebalance, rebalanceAnnotation, rebalanceOptionsBuilder);
+                break;
+            case Rebalancing:
+                nextStateFuture = onRebalancing(apiClient, clusterRebalance, rebalanceAnnotation);
+                break;
+            case Stopped:
+                nextStateFuture = onStop(apiClient, rebalanceAnnotation, rebalanceOptionsBuilder);
+                break;
+            case Ready:
+                nextStateFuture = Future.succeededFuture(clusterRebalance.getStatus());
+                break;
+            case NotReady:
+                nextStateFuture = Future.failedFuture(new RuntimeException(clusterRebalance.getStatus().getConditions().get(0).getMessage()));
+                break;
+            default:
+                nextStateFuture = Future.failedFuture(new RuntimeException("Unexpected state " + currentState));
+        }
+
+        return nextStateFuture
+                .compose(clusterRebalanceDesiredStatus -> {
+                    // due to a long rebalacing operation that takes the lock for the entire period, more events related to resource modification could be
+                    // queued with a stale resource (updated by the rebalacing holding the lock), so when we need to get the current updated resource
+                    return clusterRebalanceOperator.getAsync(reconciliation.namespace(), reconciliation.name())
+                            .compose(fetchedClusterRebalance -> {
+                                return updateStatus(fetchedClusterRebalance, clusterRebalanceDesiredStatus, clusterRebalanceOperator,
+                                        clusterRebalanceDesiredStatus.getConditions().get(0).getType(), null)
+                                        .compose(c -> {
+                                            log.info("{}: State updated to [{}] with annotation {}={} ",
+                                                    reconciliation, c.getStatus().getConditions().get(0).getType(), ANNO_STRIMZI_IO_REBALANCE,
+                                                    c.getMetadata().getAnnotations().get(ANNO_STRIMZI_IO_REBALANCE));
```

Comment:
I would factor this `c.getMetadata().getAnnotations().get(ANNO_STRIMZI_IO_REBALANCE)` into a `rebalanceAnnotation(KafkaClusterRebalance)` method. It will make the code a lot easier to read.